### ISSUE TITLE
Local Parameterization treatment of Jacobians is inefficient

### ISSUE THREAD

**Issue by dellaert** [ceres-solver/ceres-solver#615]
TL;dr: one could save a bundle of compute if one could just fill the `m*d` Jacobians, not the  `m*D` Jacobians.

Long: I am doing some optimization over SO(n), with `n>= 2`. The ambient dimension `D` is `n*n`, e.g. 64 for `n=8`, but the local dimension `d` is much smaller: `n*(n-1)/2`, i.e. for `n=8` it is only 28 << 64. 

However, the CostFunction::Evaluate is asked to compute `m*64` Jacobians, which is then multiplied in ResidualBlock by a `64*28` matrix to implement the chain rule. In my application `m` is also `n^2`, and *I know* how to compute the `m*d` Jacobians. Instead, each call to EvaluateWithJacobians now incurs a large copy cost and a large `O(m*D*D)` multiplication cost.

I paste the resulting profile (on MAcOS, Ceres 2.0rc) below for `n=8`, and as you can see almost 80% is spent on the chain rule:
![image](https://user-images.githubusercontent.com/10515273/89828583-3ce3f000-db27-11ea-8675-9aa826c19355.png)

Issue: if we had a version of Evaluate that expects one to fill the `m*d` Jacobians, local parameterizations could gain a nice performance boost.

Best
Frank

PS in my application there are also large parts of the `n*d` Jacobian that are zero, so it's doubly painful to see all those multiply by zero calls dominate `ResidualBlock::Evaluate`.

**Comment by sandwichmaker** [comment]
@dellaert you are right and this is a wart in the ceres API, that I haven't quite figured out how to fix in a proper fashion. Manifold support came to ceres after the CostFunction API was designed . And because we allow the user to change the manifold without paying attention to the CostFunction (things like subsetparameterization), making the costfunction aware is problematic from an API perspective.

That said, I think there maybe a fix here, which is that LocalParameterization supports a Multiply by Jacobian method for precisely this sort of thing, and perhaps we should be using that, which can then utilize the special structure of the parameterization to do the multiplication fast rather than doing it using generic multiplication.

**Comment by sandwichmaker** [comment]
Try this, in

residual_block.cc
replace

```
// Update the jacobians with the local parameterizations.
  if (jacobians != nullptr) {
    for (int i = 0; i < num_parameter_blocks; ++i) {
      if (jacobians[i] != nullptr) {
        const ParameterBlock* parameter_block = parameter_blocks_[i];

        // Apply local reparameterization to the jacobians.
        if (parameter_block->LocalParameterizationJacobian() != nullptr) {
          // jacobians[i] = global_jacobians[i] * global_to_local_jacobian.
          MatrixMatrixMultiply<Dynamic, Dynamic, Dynamic, Dynamic, 0>(
              global_jacobians[i],
              num_residuals,
              parameter_block->Size(),
              parameter_block->LocalParameterizationJacobian(),
              parameter_block->Size(),
              parameter_block->LocalSize(),
              jacobians[i], 0, 0,  num_residuals, parameter_block->LocalSize());
        }
      }
    }
  }
```

with

```
// Update the jacobians with the local parameterizations.
  if (jacobians != nullptr) {
    for (int i = 0; i < num_parameter_blocks; ++i) {
      if (jacobians[i] == nullptr) {
        continue;
      }

      const ParameterBlock* parameter_block = parameter_blocks_[i];
      const LocalParameterization* local_param =
          parameter_block->local_parameterization();
      if (local_param == nullptr) {
        continue;
      }

      local_param->MultiplyByJacobian(parameter_blocks_[i]->state(),
                                      num_residuals, global_jacobians[i],
                                      jacobians[i]);
    }
  }
```

**Comment by sandwichmaker** [comment]
and actually the slowdown is for two reasons.

1. We cache the local parameterization jacobians so they do not have to be computed over and over again.
2. And generally speaking we expect them to be small, so the matrix multiplication routines are optimized for that and should work.

So, another thing to try here will be to replace the call to  `MatrixMatrixMultiply` with  `MatrixMatrixMultiplyEigen`, which will use Eigen rather than our small BLAS routines to do the multiplication, which should be considerably more efficient.

**Comment by sandwichmaker** [comment]
I would recommend the second suggestion before the first, since that would require you to implement an efficent MultibyByJacobian routine, whereas you should be able to make a oneline change to ceres and see how it does for you.

**Comment by dellaert** [comment]
I actually like suggestion (1) quite a bit, and in fact could that not be done *always* ?? You'd just have to make sure the default  implementation does the right thing (i.e., the code that was replaced).

Now, it is also an opportunity to solve the entire thing, because if we get to overwrite MultiplyByJacobian, then could I not opt to not multiply *at all* ? I'll just compute the `m*d` jacobian there :-)

**Comment by dellaert** [comment]
It's a bit hacky. I still think an API change that has an optional method to compute m*d jacobians would be the best way to go. It would indeed need some though on how to integrate with subset etc (which I don't know much about).

**Comment by sandwichmaker** [comment]
could you try the two options and let me know which works better?
The reason to not do it always is that if you have a lot of residual blocks depending on the same parameter block, then it maybe the case (at least this is the assumption right now), that computing and caching the local parameterization Jacobian is the efficient thing to do.

**Comment by sandwichmaker** [comment]
I suspect that suggestion 2 will give you a substantial performance boost.

**Comment by dellaert** [comment]
> I suspect that suggestion 2 will give you a substantial performance boost.

Not so much. 77% -> 71% of Evaluate. Looking into other thing...

**Comment by dellaert** [comment]
Suggestion 1 is 20% *more* expensive. The special structure - and there is one - is mostly present in the m*D matrix, and is unavailable in MultiplyByJacobian. 't was a nice thought.

**Comment by sandwichmaker** [comment]
@dellaert so you have some truly large residual blocks, both in height and in width and they are breaking some of the assumptions we have made. Size 64 parameter blocks are quite rare for us.  BTW are you using autodiff per chance?

**Comment by dellaert** [comment]
No autodiff :-)

**Comment by dellaert** [comment]
I found a *bit* of special structure, but it does not make a big change. m*d is the way to go here (as in GTSAM :-)

**Comment by sandwichmaker** [comment]
Let me think about this a bit more.

**Comment by dellaert** [comment]
Parting thought for today: things would already be much better if the *cost function* could implement the chain rule. For example, a `FrobeniusPrior` factor I made on SO(n) has the `n^2 x n^2` identity matrix as its "global" Jacobian. So, the hypothesized MultiplyLeft() would be simple be a null operation, returning the `D*d' matrix passed in. That way, the local parameterization Jacobians can still be cached and be reused across all factors associated with it.

**Comment by sandwichmaker** [comment]
Interesting idea.
One question for you. What sized problem are you solving? How many
rotations and how many pairwise terms ?

**Comment by sandwichmaker** [comment]
The reason I ask is that got most problem s the linear solver takes way
longer than evaluation. Is there a Summary::FullReport that you can share?

**Comment by dellaert** [comment]
Sure. This is part of Shonan Averaging, of course :-) I can even add you to the evolving repo.
Here is n=8:

```
Number of poses: 1661
Number of constraints: 6275
iter      cost      cost_change  |gradient|   |step|    tr_ratio  tr_radius  ls_iter  iter_time  total_time
0  3.235031e+00    0.00e+00    8.43e-01   0.00e+00   0.00e+00  1.00e+04        0    3.26e-01    3.87e-01
1  1.330446e-03    3.23e+00    1.14e-04   1.00e+00   1.00e+00  3.00e+04        1    1.16e+00    1.54e+00
2  1.307784e-03    2.27e-05    3.92e-06   1.63e-01   1.00e+00  9.00e+04        1    9.57e-01    2.50e+00
3  1.301641e-03    6.14e-06    2.21e-06   2.38e-01   1.00e+00  2.70e+05        1    9.50e-01    3.45e+00
4  1.294915e-03    6.73e-06    1.18e-06   3.97e-01   1.01e+00  8.10e+05        1    9.67e-01    4.42e+00
5  1.292127e-03    2.79e-06    4.25e-07   3.52e-01   1.01e+00  2.43e+06        1    9.69e-01    5.39e+00
6  1.291845e-03    2.82e-07    8.35e-08   1.33e-01   1.02e+00  7.29e+06        1    9.59e-01    6.35e+00
7  1.291839e-03    5.60e-09    8.73e-09   2.03e-02   1.02e+00  2.19e+07        1    9.56e-01    7.30e+00
Solver Summary (v 2.0.0-eigen-(3.3.7)-lapack-suitesparse-(5.8.1)-cxsparse-(3.2.0)-acceleratesparse-eigensparse-no_openmp)
Original                  Reduced
Parameter blocks                         1661                     1661
Parameters                             106304                   106304
Effective parameters                    46508                    46508
Residual blocks                          6276                     6276
Residuals                              150664                   150664
Minimizer                        TRUST_REGION
Sparse linear algebra library    SUITE_SPARSE
Trust region strategy     LEVENBERG_MARQUARDT
Given                     Used
Linear solver          SPARSE_NORMAL_CHOLESKY   SPARSE_NORMAL_CHOLESKY
Threads                                     1                        1
Linear solver ordering              AUTOMATIC                     1661
Cost:
Initial                          3.235031e+00
Final                            1.291839e-03
Change                           3.233740e+00
Minimizer iterations                        8
Successful steps                            8
Unsuccessful steps                          0
Time (in seconds):
Preprocessor                         0.061163
Residual only evaluation           0.469360 (8)
Jacobian & residual evaluation     2.274173 (8)
Linear solver                      4.757974 (8)
Minimizer                            7.951506
Postprocessor                        0.058893
Total                                8.071563
Termination:                      CONVERGENCE (Function tolerance reached. |cost_change|/cost: 1.863627e-08 <= 1.000000e-06)
<End of Run>
```

And here is bottom-level, n=3, where the problems is not as severe (but linear < Jacobian eval)

```
Number of poses: 1661
Number of constraints: 6275
iter      cost      cost_change  |gradient|   |step|    tr_ratio  tr_radius  ls_iter  iter_time  total_time
0  3.235031e+00    0.00e+00    8.43e-01   0.00e+00   0.00e+00  1.00e+04        0    2.01e-02    2.93e-02
1  1.330446e-03    3.23e+00    1.14e-04   1.00e+00   1.00e+00  3.00e+04        1    4.05e-02    6.99e-02
2  1.307784e-03    2.27e-05    3.92e-06   1.63e-01   1.00e+00  9.00e+04        1    3.16e-02    1.02e-01
3  1.301641e-03    6.14e-06    2.21e-06   2.38e-01   1.00e+00  2.70e+05        1    3.17e-02    1.33e-01
4  1.294915e-03    6.73e-06    1.18e-06   3.97e-01   1.01e+00  8.10e+05        1    3.11e-02    1.64e-01
5  1.292127e-03    2.79e-06    4.25e-07   3.52e-01   1.01e+00  2.43e+06        1    3.05e-02    1.95e-01
6  1.291845e-03    2.82e-07    8.35e-08   1.33e-01   1.02e+00  7.29e+06        1    3.03e-02    2.25e-01
7  1.291839e-03    5.60e-09    8.73e-09   2.03e-02   1.02e+00  2.19e+07        1    3.09e-02    2.56e-01
Solver Summary (v 2.0.0-eigen-(3.3.7)-lapack-suitesparse-(5.8.1)-cxsparse-(3.2.0)-acceleratesparse-eigensparse-no_openmp)
Original                  Reduced
Parameter blocks                         1661                     1661
Parameters                              14949                    14949
Effective parameters                     4983                     4983
Residual blocks                          6276                     6276
Residuals                               56484                    56484
Minimizer                        TRUST_REGION
Sparse linear algebra library    SUITE_SPARSE
Trust region strategy     LEVENBERG_MARQUARDT
Given                     Used
Linear solver          SPARSE_NORMAL_CHOLESKY   SPARSE_NORMAL_CHOLESKY
Threads                                     1                        1
Linear solver ordering              AUTOMATIC                     1661
Cost:
Initial                          3.235031e+00
Final                            1.291839e-03
Change                           3.233740e+00
Minimizer iterations                        8
Successful steps                            8
Unsuccessful steps                          0
Time (in seconds):
Preprocessor                         0.009216
Residual only evaluation           0.030637 (8)
Jacobian & residual evaluation     0.077442 (8)
Linear solver                      0.050610 (8)
Minimizer                            0.263804
Postprocessor                        0.002402
Total                                0.275422
Termination:                      CONVERGENCE (Function tolerance reached. |cost_change|/cost: 1.863627e-08 <= 1.000000e-06)
<End of Run>
```

**Comment by dellaert** [comment]
n=8 is a rare case. But If possible I'd like to "flatten the curve" in n :-)

**Comment by sandwichmaker** [comment]
If I am reading this right, then the situation is worse for n = 3 than n = 8, where the jacobian evaluation is slower than the linear solve. BTW this is actually a perfect case for using threading. Another idea would be to use LBFGS since your objective function does not appear to be too hard.

**Comment by dellaert** [comment]
But adding a call to have the cost-function to the multiplication would not break the API at all, right? I think it will be useful in quite a few situations. I have some plans :-)

**Comment by dellaert** [comment]
Cost functions above are easy because of good initial estimates. Harder if starting from random.

**Comment by sandwichmaker** [comment]
I do not think I quite understand your suggestion @dellaert. Could you elaborate? and what are these plans?

**Comment by dellaert** [comment]
> I do not think I quite understand your suggestion @dellaert. Could you elaborate? and what are these plans?

Suggestion, calling J global Jacobian and C chain rule Jacobian
- Currently, ResidualBlock could do: J = Cost->Jacobian(x);  JC = LocalParam->DoChainRule(x, J)
- My suggestion: ResidualBlock does: C = LocalParam(x);  JC = Cost->DoChainRule(x, C)

Plans: later :-)

**Comment by sandwichmaker** [comment]
Alright, let me think about it. If something like this happens it will not happen before the end of August. It may also be a good time to revamp the LocalParameterization object into a full fledged Manifold object with Plus and Minus operations, which will make things like constrained optimization a lot saner.

**Comment by dellaert** [comment]
Yes ! Or you could call them `retract` and `local_coordinates` :-)

**Comment by sandwichmaker** [comment]
Plus and minus :)

**Comment by NikolausDemmel** [comment]
Great to see some movement on the topic.

Cross referencing some previous discussion:
- https://groups.google.com/g/ceres-solver/c/ALoHTiDYE04
- https://groups.google.com/g/ceres-solver/c/fs8iNI9_F7Q
- https://github.com/ceres-solver/ceres-solver/issues/387
- https://github.com/ceres-solver/ceres-solver/issues/303

People in the wild are using the workaround of putting "identity + 0 columns" in the local parameterization (e.g. [VINS-Mono](https://github.com/HKUST-Aerial-Robotics/VINS-Mono/issues/82#issuecomment-544667450)), but that only allows to use the `m*d` Jacobian, but still performs the multiplication. It also breaks things like automatic jacobian checker...

> My suggestion: ResidualBlock does: C = LocalParam(x); JC = Cost->DoChainRule(x, C)

The nice thing about something like this would be that the same local parameterization can be used with residuals that use the default implementation (i.e. multiply `C` with the `m*D` Jacobian) and residuals that already compute the `m*d` directly, where `DoChainRule` is effectively a no-op. I wonder if there is a way to skip the computation of `C` alltogether if it is not used, but that is a very minor optimization.

> The reason to not do it always is that if you have a lot of residual blocks depending on the same parameter block, then it maybe the case (at least this is the assumption right now), that computing and caching the local parameterization Jacobian is the efficient thing to do.

I would assume this is true if you internally use the chain rule for the `m*d` Jacobian. For Lie groups in my experience it's often much easier and more compact to directly derive the Jacobian w.r.t. the Lie algebra increment analytically. A simple example is rotating a 3d point with SO(3):

```
d/de R exp(e) p = d/de R (I + e^) p = d/de R e^ p = d/de - Rp^ e = -Rp^,
``` 

where `R` is a rotation matrix, `e` is the R^3 lie algebra increment, `^` is the "hat" operator and `p` is the 3d point.

Maybe one very common scenario in ceres where what you say does apply, is for autodiff residuals. If you would use autodiff all the way to the increment in every residual block, you'd effectively do the chain rule many times and would likely benefit from the caching. So I agree that it might make sense if this would stay the default behaviour, and the alternative be opt-in for people that wanna supply the analytic `m*d` Jacobian in the cost function.

(It's not central to this change, but I wonder if there is a way to teach auto-diff how to derive the `m*d` efficiently w/o chain rule (for the retraction part), at least for special cases where we know that the local parameterization implements a proper Lie group retraction. Maybe not, at least as far as I understand how Jet works...)

**Comment by NikolausDemmel** [comment]
> > My suggestion: ResidualBlock does: C = LocalParam(x); JC = Cost->DoChainRule(x, C)
> 
> The nice thing about something like this would be that the same local parameterization can be used with residuals that use the default implementation (i.e. multiply `C` with the `m*D` Jacobian) and residuals that already compute the `m*d` directly, where `DoChainRule` is effectively a no-op. I wonder if there is a way to skip the computation of `C` alltogether if it is not used, but that is a very minor optimization.

One more thought: What is nice about the way ceres works at the moment is that the cost function is completely agnostic of which local parameterization is used and can be combined with different choices. In the past, I found it quite convenient to easily switch local parameterizations for experimentation, or use different local parameterizations for different parameter blocks of the same "type". I think this is an advantage over other libraries where parameter blocks are typed (as in static C++ types) and the retraction is tied to the type (but of course, typed parameter blocks also have advantages).

With `C = LocalParam(x); JC = Cost->DoChainRule(x, C)` this would not be true any more. The cost function has to know which retraction is used to compute the correct `m*d` Jacobian. Of course you could just say it's the programmer's responsibility to make sure they are compatible, and maybe that is ok given that it's the special and not the general case in ceres. But I think it would be great if there was a way that the cost function can know which local parameterization it is combined with, to at least error out, or if the user wants, implement conditional code that supports multiple cases.

**Comment by dellaert** [comment]
Hey @NikolausDemmel : 
- Yes on Lie groups: in GTSAM we only ever compute the m*d Jacobians. In many cases this is through chain rules or our "expressions" mini-DSL that does reverse AD, but even in that case we rarely if ever have inner dimensions of size D.
- On the last point I disagree: the nice thing is exactly that the local parameterization can be changed as we can just feed in a *different* `C` into `Cost->DoChainRule(x, C)`. For example, in my FrobeniusPrior example we just pass C back to the caller, *whatever* C is.

**Comment by dellaert** [comment]
Oh, but I just saw the issue with that: `double*`. @sandwichmaker I have to say that `double*` is a curse in using Ceres :-) It's been a while since I have had to step into gdb/lldb to figure out issues in my code! IMHO this can all be solved rather easily by using Eigen::Ref as the API, as well. Maybe this is not acceptable in general, but for an advanced API like  `DoChainRule(x, C)` it could be done, so that that this method at least knows the dimension of the block of memory passed in. That would solve issues for example with subset parameterizations as well, I think?

**Comment by NikolausDemmel** [comment]
Hi @dellaert,

> On the last point I disagree: the nice thing is exactly that the local parameterization can be changed as we can just feed in a different C into Cost->DoChainRule(x, C). For example, in my FrobeniusPrior example we just pass C back to the caller, whatever C is.

Ok, got it. Yes, in your example of `FrobeniusPrior`, it would indeed be as you say. I didn't realize you were talking about this case. But since the "FrobeniusPrior" is pretty much the only cost function that can return C unchanged, the only thing that a custom `DoChainRule` implementation could do with C, is exploiting some sparsity in `J`, when multiplying `JC`. Nothing about `C` can be assumed if you want it to work with any local parameterization.

So what I was thinking about and commenting on, is the possibility that a cost function that overrides `DoChainRule` can effectively ignore the passed in `C`, if it already computed the `m*d` Jacobian in `Evaluate`, and simply return the stored Jacobian. But then, as mentioned, we do in fact have to assume a specific local parameterization and my comment was if / how it could be made known to the cost function, which local parameterization is actually set, to either use conditional code, or at least error out.

Side note: I think it cannot work exactly that way in any case, since the cost-function's `Evaluate` API is supposed to be stateless for multithreading AFAIK.

To sum up, I think it's good to keep these two slightly different use cases in mind for any API change:
 1. You want to use the `D*d` Jacobian `C` from the local parameterization, but the cost function wants to control how it's multiplied with the `m*D` Jacobian `J` (e.g. "not at all", if J is identity).
 2. You don't want to use `C` at all, since the cost function provides `m*d` directly, without going at all via the `D` dimension.

**Comment by dellaert** [comment]
True, but exploiting sparsity is *big * :-) I think Frobenius is only the simplest case, but the Shonan factors are just as sparse, just different patterns than identity (basically the J is a set of identity and zero blocks).

**Comment by sandwichmaker** [comment]
Going to close this, as currently there is no work planned on this.
